src/System.Runtime.Numerics/src/System/Numerics/Complex.cs

```diff
@@ -320,6 +372,117 @@ public static Complex Atan(Complex value)
+            // get the quadratic equation t^2 - 2 i z t - 1 = 0. The solution is t = i z + sqrt(1 - z^2), so
+            //   w = arcsin(z) = - i log( i z + sqrt(1 - z^2) )
+            // Decompose z = x + i y, multiply out i z + sqrt(1 - z^2), use log(s) = |s| + i arg(s), and do a
+            // bunch of algebra to get the components of w = arcsin(z) = u + i v
+            //   u = arcsin(\beta)  v = sign(y) log(\alpha + sqrt(\alpha^2 - 1))
+            // where
+            //   \alpha = (\rho + \sigma)/2  \beta = (\rho - \sigma)/2
+            //   \rho = sqrt((x + 1)^2 + y^2)  \sigma = \sqrt((x - 1)^2 + y^2)
+            // This appears in DLMF section 4.23. (http://dlmf.nist.gov/4.23), along with the analogous
+            //   arccos(w) = arccos(\beta) - i sign(y) log(\alpha + sqrt(\alpha^2 - 1))
+            // So \alpha and \beta together give us arcsin(w) and arccos(w).
+
+            // As written, \alpha is not susceptible to cancelation errors, but \beta is. To avoid cancelation, note
+            //   \beta = (\rho^2 - \sigma^2)/(\rho + \sigma))/2 = (2 x)/(\rho + \sigma) = x / \alpha
+            // which is not subject to cancelation. Note \alpha >= 1 and |\beta| <= 1.
+
+            // For \alpha ~ 1, the argument of the log is near unity, so we compute (\alpha - 1) instead,
+            // and write the argument as 1 + (\alpha - 1) + \sqrt{(\alpha - 1)(\alpha + 1)}.
+            // For \beta ~ 1, arccos does not accurately resolve small angles, so we compute the tangent of the angle
+            // instead. Hull and Tang derive formulas for (\alpha - 1) and \beta' = \tan(u) that do not suffer
+            // cancelation for these cases.
+
+            // For simplicity, we assume all positive inputs and return all positive outputs. The caller should
+            // assign signs appropriate to the desired cut conventions. We return v directly since its magnitude
+            // is the same for both arcsin and arccos. Instead of u, we usually return \beta and sometimes \beta'.
+            // If \beta' is not computed, it is set to -1; if it is computed, it should be used instead of \beta
+            // to determine u. Compute u = \arcsin(\beta) or u = \arctan(\beta') for arcsin, u = \arccos(\beta)
+            // or \arctan(1/\beta') for arccos.
+
+            // For x or y large enough to overflow \alpha^2, we can simplify our formulas and avoid overflow.
+            if ((x > s_asinOverflowThreshold) || (y > s_asinOverflowThreshold))
+            {
+
+                b = -1.0;
+                bPrime = x / y;
+
+                double small, big;
+                if (x < y)
+                {
+                    small = x;
+                    big = y;
+                }
+                else
+                {
+                    small = y;
+                    big = x;
+                }
+                double ratio = small / big;
+                v = Math.Log(2.0) + Math.Log(big) + 0.5 * Log1P(ratio * ratio);
```

Comment:
C compilers compute the most basic math functions at compile time. The C# compiler doesn't, but I bet the JIT-ter does it at JIT time, in which case it's only done once per run.

At some point we should create an internal static Constants class containing mathematical constants to be used in algorithms throughout the library.

For this PR, though, I'm putting this in a static readonly field as you requested.